langchain4j-agentic/src/main/java/dev/langchain4j/agentic/cognisphere/Cognisphere.java

```diff
@@ -0,0 +1,233 @@
+        } else if (kind == Kind.PERSISTENT) {
+            flush();
+        }
+        CognisphereRegistry.resetCurrentAgentId();
+    }
+
+    private void flush() {
+        lock.writeLock().lock();
+        try {
+            CognisphereRegistry.update(this);
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private void registerContext(AgentSpecification agentSpecification, Object agent, Object[] input, Object response) {
+        if (agent instanceof ChatMemoryAccess agentWithMemory) {
+            ChatMemory chatMemory = agentWithMemory.getChatMemory(key.id());
+            if (chatMemory != null) {
+                List<ChatMessage> agentMessages = chatMemory.messages();
+                for (int i = agentMessages.size() - 1; i >= 0; i--) {
+                    if (agentMessages.get(i) instanceof UserMessage userMessage) {
+                        // Only add to the cognisphere's context the last UserMessage ...
+                        context.add(new AgentMessage(agentSpecification.name(), userMessage));
+                        // ... and last AiMessage response, all other messages are local to the invoked agent internals
+                        context.add(new AgentMessage(agentSpecification.name(), agentMessages.get(agentMessages.size() - 1)));
+                        return;
+                    }
+                }
+            }
+        }
+        if (agentSpecification.description() != null && !agentSpecification.description().isBlank()) {
+            context.add(new AgentMessage(agentSpecification.name(), new UserMessage(agentSpecification.description())));
+            context.add(new AgentMessage(agentSpecification.name(), AiMessage.aiMessage(response.toString())));
+        }
+    }
+
+    public List<AgentMessage> context() {
+        return context;
+    }
+
+    public String contextAsConversation(String... agentNames) {
+        Predicate<String> agentFilter = agentNames != null && agentNames.length > 0 ?
+                List.of(agentNames)::contains :
+                agent -> true;
+
+        StringBuilder sb = new StringBuilder();
+        for (AgentMessage agentMessage : context) {
+            if (!agentFilter.test(agentMessage.agentName())) {
+                continue;
+            }
+            ChatMessage message = agentMessage.message();
+            if (message instanceof UserMessage userMessage) {
+                sb.append("User: ").append(userMessage.singleText()).append("\n");
+            } else if (message instanceof AiMessage aiMessage) {
+                sb.append(agentMessage.agentName()).append(" agent: ").append(aiMessage.text()).append("\n");
+            }
+        }
+
+        String contextAsConversation = sb.toString();
+        LOG.info("Cognisphere context as conversation: '{}'", contextAsConversation);
+        return contextAsConversation;
+    }
+
+    public List<AgentCall> getAgentInvocations(String agentName) {
```

Comment:
```suggestion
    public List<AgentCall> agentCalls(String agentName) {
```
To be consistent with `AgentCall`.

There are many places in this module that refer to "invocation", but the name of the class is "call", can we make it consistent please?

Reply:
IMHO "invocation" seems more appropriate than "call", so perhaps rename `AgentCall` to `AgentInvocation`?

Reply:
I fixed the method name.

> IMHO "invocation" seems more appropriate than "call", so perhaps rename `AgentCall` to `AgentInvocation`?

Agreed, I used `AgentCall` only to avoid a name clash with the `AgentInvocation` class that I have in the `supervisor` package. Anyway this class shouldn't be part of the public API and I'm accessing it only to check its content in tests.